src/thorin/transform/resolve_loads.cpp

```diff
@@ -0,0 +1,189 @@
+}
+
+static const Def* try_resolve_load(DefMap<bool>& safe_slots, const Def* def, const Load* target_load, const Def* slot, size_t depth) {
+    auto& world = def->world();
+    while (true) {
+        auto mem_op = def->isa<MemOp>();
+        if (!mem_op)
+            return nullptr;
+
+        const Def* parent = nullptr;
+        auto load  = Load::is_out_mem(mem_op->mem());
+        auto store = mem_op->mem()->isa<Store>();
+        if (load || store) {
+            auto parent_ptr   = load ? load->ptr() : store->ptr();
+            auto parent_slot  = find_slot(safe_slots, parent_ptr);
+            auto parent_depth = parent_slot.second;
+            if (parent_slot.first == slot && is_pointer_prefix(target_load->ptr(), depth, parent_ptr, parent_slot.second)) {
+                if (depth >= parent_depth) {
+                    const Def* value = load ? world.extract(load, world.literal_qs32(1, Debug())) : store->val();
+                    return extract_from_leas(value, target_load->ptr(), depth - parent_depth);
+                } else if (store) {
+                    const Def* value = try_resolve_load(safe_slots, store, target_load, slot, depth);
+                    if (!value)
+                        return nullptr;
+                    return insert_from_leas(value, store->ptr(), store->val(), parent_depth - depth);
+                }
+            }
+            parent = load ? load->as<Def>() : store;
+        } else if (auto enter = Enter::is_out_mem(mem_op->mem())) {
+            // TODO: Solve undefined allocs (need to inspect the frame)
+            parent = enter;
+        } else {
+            return nullptr;
+        }
+
+        def = parent;
+    }
+}
+
+static bool resolve_loads(DefMap<bool>& safe_slots, const Scope& scope) {
+    auto& world = scope.world();
+    bool todo = false;
+    for (const auto& block : schedule(scope, Schedule::Late)) {
+        for (auto primop : block) {
+            // Compute whether this slot is safe to transform
+            auto slot = primop->isa<Slot>();
+            if (slot) {
+                is_safe_slot(safe_slots, slot);
+                continue;
+            }
+
+            auto load = primop->isa<Load>();
+            if (!load)
+                continue;
+            // Reolve loads that are from a safe slot
+            auto slot_depth = find_slot(safe_slots, load->ptr());
+            if (!slot_depth.first)
+                continue;
+            if (auto value = try_resolve_load(safe_slots, load, load, slot_depth.first, slot_depth.second)) {
+                load->replace(world.tuple({ load->mem(), value }, load->debug()));
+                todo = true;
+            }
+        }
+    }
+    return todo;
+}
+
+bool resolve_loads(World& world) {
+    DefMap<bool> safe_slots;
+    bool todo = false;
+    Scope::for_each(world, [&] (const Scope& scope) {
+        todo |= resolve_loads(safe_slots, scope);
+    });
```

Comment:
Since you are changing stuff within the scope it might be necessary to invoke `scope.update()`.